src/comparator/wrappers.py

```diff
@@ -770,6 +756,58 @@ def _get_api_version(
         package = self.file_set_pb.file[0].package
         return re.search(version, package).group()
 
+    def _get_source_code_locations_map(
+        self, file_set: descriptor_pb2.FileDescriptorSet
+    ) -> Dict[str, Dict[Tuple[int, ...], descriptor_pb2.SourceCodeInfo.Location]]:
+        source_code_locations_map = {}
+        for fd in file_set.file:
+            # Iterate over the source_code_info and place it into a dictionary.
+            #
+            # The comments in protocol buffers are sorted by a concept called
+            # the "path", which is a sequence of integers described in more
+            # detail below; this code simply shifts from a list to a dict,
+            # with tuples of paths as the dictionary keys.
+            source_code_locations: Dict[
+                Tuple[int, ...], descriptor_pb2.SourceCodeInfo.Location
+            ] = {}
+            for location in fd.source_code_info.location:
+                source_code_locations[tuple(location.path)] = location
+            source_code_locations_map[fd.name] = source_code_locations
+        return source_code_locations_map
+
+    def _get_resource_database(
+        self,
+        file_set: descriptor_pb2.FileDescriptorSet,
+        source_code_locations_map: Dict[
+            str, Dict[Tuple[int, ...], descriptor_pb2.SourceCodeInfo.Location]
+        ],
+    ):
+        resources_database = ResourceDatabase()
+        for fd in file_set.file:
+            source_code_locations = source_code_locations_map[fd.name]
+            # File-level resource definition.
+            for i, resource in enumerate(
+                fd.options.Extensions[resource_pb2.resource_definition]
+            ):
+                resource_path = (8, 1053, i)
```

Comment:
resource_path means: 8-> file options, 1053-> resource_definition, i-> index of the resource.

Reply:
Sure, same below. Thanks